### DISCUSSION TITLE
eslint configuration with absolute imports in Next 9.4

### DISCUSSION THREAD

**Discussion by conanbatt** [vercel/next.js#13048] [Help] [answered]
The absolute imports project doesn't have an eslint configuration. How to setup eslint to recognize absolute imports with nextjs resolvers?

9.4 documentation recommends not adding resolves to webpack and use the jsconfig.js file. Is the rule not supported without adding libraries? How to approach?

**Comment by rafaelalmeidatk** [comment, marked as answer]
You need to explain to eslint how it should find your files. First, add the [`eslint-plugin-import`](https://www.npmjs.com/package/eslint-plugin-import) package to your project. If you are using TypeScript, add the [`eslint-import-resolver-typescript`](https://github.com/alexgorbatchev/eslint-import-resolver-typescript) package and it will be able to find the files.

If you are using JS, add the [`eslint-import-resolver-alias`](https://www.npmjs.com/package/eslint-import-resolver-alias) package and write the path aliases. This will depends on your aliases, here is an example config:

```json5
{
  // your other eslint configs ...
  // "rules": { ... },
  "settings": { 
    "import/resolver": {
      "alias": {
        "map": [
          ["components", "./components"],
          ["lib", "./lib"]
        ]
      }
    }
  }
}
```

**Reply by conanbatt** [reply]
Amazing! Thanks

(I did have to take out the ~ to make it work).

**Comment by ryanrouleau** [comment]
For anyone that comes across and is not familiar with `eslint-import-resolver-typescript` you have to add

```json
  "settings": {
    "import/resolver": {
      "typescript": {
        "directory": "."
      }
    }
  },
```

to your `.eslintrc.json` where the value of `"directory"` is the relative path to the directory holding your `tsconfig.json`.

In my case both my `.eslintrc.json` and `tsconfig.json` are in the same directory so the val of `"directory"` is `"."`.

Hopefully this saves you the 4 min from wondering why it's not working then looking at the docs of `eslint-import-resolver-typescript` lol

**Reply by anku255** [reply]
For anyone coming across this, the `directory` key has been renamed to `project`. So it becomes

```
  "settings": {
    "import/resolver": {
      "typescript": {
        "project": "."
      }
    }
  },
```

**Reply by michielvanderros-tomtom** [reply]
@anku255's tip is gold.
Maybe good to know for anyone with a monorepo who stumbled upon this post:
If you have a root .eslintrc file and a nested package that is a NextJS project with aliases like `components/Layout`, eslint may flag your imports as errors. When you add another .eslintrc file to your nested package, it may start ignoring the root rules for VSCode ESLint/Prettier.
But when you add the following to your package in a new .eslintrc file it should both work (auto formatting and all NextJS aliases for short imports without relative paths like `../../../components`)
```
{
  "extends": "../../../.eslintrc", // May not be striclty necessary, but root .eslintrc rules are applied to YourProject now.
  "settings": {
    "import/resolver": {
      "typescript": {
        "project": "packages/apps/your-project"
      }
    }
  }
}

**Comment by adriandmitroca** [comment]
Just in case you ended up here and for some reason TypeScript solution isn't good fit for you - here's basic configuration that you need for JavaScript:

```json
"settings": {
  "import/resolver": {
    "node": {
      "paths": [
        "."
      ]
    }
  }
},
```

**Reply by vicasas** [reply]
This is not worked whit this

```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@/components/*": ["components/*"]
    }
  }
}
```

**Reply by lucasmarinb** [reply]
In my case, for the absolute imports it was necessary to add a `node` entry, and for the module path alias i needed to add `jsx` in the `extensions` entry:
```json
  "settings": { 
    "import/resolver": {
      "alias": {
        "map": [
          ["@/components", "./components"]
        ],
        "extensions": [".ts", ".js", ".jsx", ".json"]
      },
      "node": {
        "paths": ["./"]
      }
    }
  }
```

**Comment by hjaimee** [comment]
In Typescript how can I make it work when using custom paths such as `@components` . At the moment eslint is giving me the following error

`Cannot find module '@components/folder' or its corresponding type declarations.`

I have  `eslint-import-resolver-typescript` installed

**Reply by azedo** [reply]
If you have the aliases defined in the `tsconfig.json` file, it should work when you add the settings for `import/resolver` in the `eslintrc` file, like @ryanrouleau said above -> https://github.com/vercel/next.js/discussions/13048#discussioncomment-17688

```json
"settings": {
  "import/resolver": {
    "typescript": {
      "directory": "."
    }
  }
},
```
